hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestRegionServerReportForDuty.java

```diff
@@ -241,6 +243,125 @@ public void run() {
+    assertEquals(1, master.getMaster().getServerManager().getDrainingServersList().size());
+    assertEquals(1, master.getMaster().getServerManager().getOnlineServers().size());
+    assertEquals(rs.getRegionServer().getServerName().toString(),
+      decommissionedServer.getServerName());
+
+    // Create a second region server
+    cluster.getConfiguration().set(HConstants.REGION_SERVER_IMPL, MyRegionServer.class.getName());
+    rs2 = cluster.addRegionServer();
+    rs2.start();
+    waitForSecondRsStarted();
+
+    // Assert that the number of decommissioned and live hosts didn't change and that the hostname
+    // of rs2 matches that of the decommissioned server
+    String rs2HostName = rs2.getRegionServer().getServerName().getHostname();
+    assertEquals(1, master.getMaster().listDecommissionedRegionServers().size());
+    assertEquals(1, master.getMaster().getServerManager().getDrainingServersList().size());
+    assertEquals(1, master.getMaster().getLiveRegionServers().size());
+    assertEquals(1, master.getMaster().getServerManager().getOnlineServers().size());
+    assertEquals(rs2HostName, decommissionedServer.getHostname());
+  }
+
+  /**
+   * Tests that the RegionServer's reportForDuty gets accepted by the master when the master is not
+   * configured to reject decommissioned hosts, even when there is a match for the joining
+   * RegionServer in the list of decommissioned servers. Test case for HBASE-28342.
+   */
+  @Test
+  public void testReportForDutyGetsAcceptedByMasterWhenNotConfiguredToRejectDecommissionedHosts()
+    throws Exception {
+    // Start a master and wait for it to become the active/primary master.
+    // Use a random unique port
+    cluster.getConfiguration().setInt(HConstants.MASTER_PORT, HBaseTestingUtil.randomFreePort());
+    cluster.getConfiguration().setInt(ServerManager.WAIT_ON_REGIONSERVERS_MINTOSTART, 1);
+    cluster.getConfiguration().setInt(ServerManager.WAIT_ON_REGIONSERVERS_MAXTOSTART, 1);
+
+    // Set the cluster to not reject decommissioned hosts (default behavior)
+    cluster.getConfiguration().setBoolean(HConstants.REJECT_DECOMMISSIONED_HOSTS_KEY, false);
+
+    master = cluster.addMaster();
+    rs = cluster.addRegionServer();
+    LOG.debug("Starting master: " + master.getMaster().getServerName());
+    master.start();
+    rs.start();
+
+    waitForClusterOnline(master);
+
+    assertEquals(0, master.getMaster().listDecommissionedRegionServers().size());
+    assertEquals(0, master.getMaster().getServerManager().getDrainingServersList().size());
+    assertEquals(1, master.getMaster().getServerManager().getOnlineServers().size());
+
+    // Decommission the region server and tries to re-add it
+    List<ServerName> serversToDecommission = new ArrayList<ServerName>();
+    serversToDecommission.add(rs.getRegionServer().getServerName());
+    master.getMaster().decommissionRegionServers(serversToDecommission, true);
```

Comment:
Is this method effectively asynchronous? Do you need to wait for some side-effect before the test can proceed?

Reply:
It is asynchronous but I can see its results in the following assertions, it gets executed in a separate thread. Do you think we need to explicitly add a `Thread.sleep` here?

Reply:
Our test hosts tend to be over-extended. Its usually safest to include an explicit wait for a specific condition. This results in more reliable tests. We have the class `o.a.h.h.Waiter` for this purpose, along with several variants of methods named `HBaseCommonTestingUtility#waitFor`, and integration with hamcrest via the `MatcherPredicate` class.